### PR TITLE
Fuzzy matching of contact list

### PR DESCRIPTION
**Pull request by StupidCodeFactory**
**CHANGES:**
Exactly match date of birth and Fuzzy match visitors names against a prisoner's
contact list.
The fuzzy match is done using the JaroWinkler algorythm (i.e did_you_mean gem)

### CODE REVIEW

File: app/views/prison/visits/_visitor_contact.html.erb [hunk 1]
```diff
@@ -27,23 +27,10 @@
...
-        <%= vf.check_box :banned %>
-        <%= vf.label :banned do %>
-          <%= t('.banned') %>
-        <% end %>
-      </div>
+      <%= render 'prison/visits/not_on_list', vf: vf, v: v  %>
+      <%= render 'prison/visits/banned', vf: vf, v: v  %>
     </div>
   </div>
 </li>
```
Comment (doctorpod) [inline app/views/prison/visits/_visitor_contact.html.erb]: Nice refactor 👍

File: spec/models/concerns/contact_list_matcher_behaviour_spec.rb [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    unless defined?(TestMatcher)
+      TestMatcher = Class.new do
+        include described_module
+      end
+    end
+
+    TestMatcher.new
+  end
+
+  describe '#catergory' do
```
Comment (doctorpod) [inline spec/models/concerns/contact_list_matcher_behaviour_spec.rb]: Pelling mistale

File: app/models/concerns/contact_list_matcher_behaviour.rb [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  end
+
+  def contacts_with_data
+    @contacts_with_data ||=
+      begin
+        if contacts.empty?
+          [no_match]
+        else
+          contacts.map do |contact|
+            [contact, {  data: { contact: contact } }]
```
Comment (alan) [inline app/models/concerns/contact_list_matcher_behaviour.rb]: 2 spaces in the hash

File: app/models/contact_list_matcher.rb [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    end
+  end
+
+  def match_date_of_birth?(contact)
+    visitor.date_of_birth.to_date == contact.date_of_birth.to_date
+  end
+
+  # rubocop:disable Metrics/MethodLength
+  def matched_bucket_for(candidate)
+    visitor_full_name   = "#{visitor.first_name} #{visitor.last_name}".downcase
```
Comment (alan) [inline app/models/contact_list_matcher.rb]: There is a method on visitor to get the full name, can we use that?

File: app/models/contact_list_matcher.rb [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        others.add(0, contact)
+      end
+    end
+  end
+
+  def match_date_of_birth?(contact)
+    visitor.date_of_birth.to_date == contact.date_of_birth.to_date
+  end
+
+  # rubocop:disable Metrics/MethodLength
```
Comment (alan) [inline app/models/contact_list_matcher.rb]: We can get rid of this rubocop by pushing some of the logic into a private method

**Comment by chrishanes** [comment]
@alan I'm not sure the select option should be pre-selected

**Comment by kyliehavelock** [comment]
Approved - good to go! @StupidCodeFactory

**Pull request merged**